### PR TITLE
[CR] Fences Project (Part 1)

### PR DESCRIPTION
**Pull request by Lap**
While hiding out in a public works area I was spent a lot of time fenced in. This made me think of a couple things:
1. Monsters and NPC's are at a horrible disadvantage when it comes to chain-link fences.
2. I really wanted to electrify these fences.

I started trying to implement electric fences, which kind of morphed into this small project.
- **Certain terrain is climbable by monsters with MF_CLIMBS or NPC's.**
  - At present I have enabled this only for chain-linked fences, NPC's, feral hunters and a few other select monsters can now chase you over them.
  - Similar to the player, NPC's have a chance of slipping and will climb with speed based on dex.
  - Should still be useful with Z levels (with modification).
- **NPC's that become "stuck" next to a nonbashable terrain object will now flee instead of waiting to die.**

**To Do**

[X] Tiny and flying monsters will no longer get the terrain status effects of ROUGH, SHARP, UNSTABLE.

[X] Support for fliers moving over climbable terrain

[X] Increased damage players can take from sharp objects (Now actually possible to trigger bleed effects while walking naked through barbed wire or broken windows)

**Stashed for Another PR**

[X] Alert players when NPC's and monsters cut themselves on terrain.

[X] NPC's now take damage from rough/sharp terrain
[ ] Prevent animals from running into electrified objects

[X] New type of fence with barbed wire on top . Easy for the player to attach to normal chain link fencing and it would damage/slow those climbing over. Would likely replace fences around substations and SAI's. [Example](http://upload.wikimedia.org/wikipedia/commons/8/83/2008-08-01_No_Tresspassing_sign_at_RDU.jpg)

[X] New type of high-security fence. Essentially unclimbable. This would be used for prisons, military bases, hazardous waste sarcophagus [example](http://www.herculesfence.com/gallery/images/gallery/main/Razor%20Ribbon%20Prison%20Fence.jpg)

[ ] Allow player to turn normal chain-link fence into barbed wire chain-link fence.

[ ] Allow players to repair broken chain-link fences.

[ ] Add ELECTRIFIED as option for vehicle parts.

[ ] Create way for player to electrify rams or plating. Extra shock damage and small stun at the cost of battery power. A nonturret security solution for your vehicles!

[X] Fence cutting damage can now be different than that of other sharp terrain.

[ ] Allow spitter zombies to spit through fences

[ ] Allow damage to vary based on rough simulation of volts/amps/current.

[ ] Allow jumper cables connected to fencing to electrify them.This will likely be the most complex part as I would prefer to not have fences count as vehicles, which means changing jumper cables to allow them to connect to fencing. I imagine I would have to search for all adjacent fencing, propagating down the line and replacing the terrain with a different electrified version. This has a whole bunch of steps that can go wrong so I'll hold off and see if anyone else has ideas. I should probably talk to narc0tiq as the power sharing will probably take me hours to figure out myself. 
- **Added ELECTRIFIED as a furniture and terrain tag.** 
  - This can be used for active generators, set piece items in factories, labs, substations, traps, etc.
  - Anyone bashing these will take electrical damage and a brief stun. Moving across the tile will also cause the effect.
  - No actual new electrified terrain in this PR

### CODE REVIEW

**Comment by Snaaty** [comment]
All of those look pretty brilliant!

**Comment by narc0tiq** [comment]
:+1:

File: src/monmove.cpp [hunk 1]
```diff
@@ -824,6 +840,15 @@ int monster::attack_at(int x, int y) {
 
 int monster::move_to(int x, int y, bool force)
 {
+    //Allows climbing monsters to move on terrain with movecost <= 0
+    if (g->m.has_flag_ter_or_furn("CLIMBABLE", x, y) && has_flag(MF_CLIMBS) && g->npc_at(x, y) == -1 && g->mon_at(x, y) == -1) {
+            if (g->u.sees( *this )){
+                    add_msg(_("The %s climbs over the %1$s."), name().c_str(),
+                            g->m.has_flag_ter_or_furn("CLIMBABLE", x, y) ? g->m.tername(x, y).c_str() : g->m.furnname(x, y).c_str());
```
Comment (BevapDin) [inline src/monmove.cpp]: `g->m.has_flag_ter_or_furn("CLIMBABLE", x, y) ` is already checked 3 lines above is definitely true. Did you mean to call `has_flag_ter`?

Also: "The %s climbs over the %1$s.", why use the monster name (first parameter) as terrain parameter again? Why not just "The %s climbs over the %s"?

File: src/game.cpp [hunk 2]
```diff
@@ -7141,6 +7141,16 @@ void game::smash()
             return; // don't smash terrain if we've smashed a corpse
         }
     }
+    bool ter_or_furn = m.has_flag_ter_or_furn( "ELECTRIFIED", smashx, smashy );
+    if (ter_or_furn) {
+            body_part bp = random_body_part();
+            if(u.deal_damage( nullptr, bp, damage_instance( DT_ELECTRIC, rng( 5, 20 ) ) ).total_damage() > 0) {
+                add_msg(m_bad, _("You are shocked by the %1$s!"),
+                    ter_or_furn ? m.tername(smashx, smashy).c_str() : m.furnname(smashx, smashy).c_str());
```
Comment (BevapDin) [inline src/game.cpp]: See below, `ter_or_furn` is definitely true at this point.

File: src/monmove.cpp (hunk 1 above)
Comment (Lap) [inline src/monmove.cpp]: I should just use has_flag_ter and I need to correct that in a few other places elsewhere in this PR.

File: src/game.cpp [hunk 3]
```diff
@@ -7141,6 +7141,16 @@ void game::smash()
             return; // don't smash terrain if we've smashed a corpse
         }
     }
+    bool ter_or_furn = m.has_flag_ter_or_furn( "ELECTRIFIED", smashx, smashy );
+    if (ter_or_furn) {
+            body_part bp = random_body_part();
```
Comment (BevapDin) [inline src/game.cpp]: I usually don't smash things with my head or my torso, but that's just me (-:

You should check for the wielded (and therefor used) item (`item::conductive()`). The character should not be damaged when bashing with a suitable (non-conductive) weapon, similar to the code in `mdefense::zapback`.

File: src/monmove.cpp [hunk 4]
```diff
@@ -693,11 +698,22 @@ int monster::bash_at(int x, int y) {
...
     if( try_bash && can_bash ) {
         int bashskill = group_bash_skill( point(x, y) );
-
+        bool ter_or_furn = g->m.has_flag_ter_or_furn("ELECTRIFIED", x, y);
+        if (ter_or_furn && not has_flag(MF_ELECTRIC)) {
+            if(deal_damage( nullptr, bp_torso, damage_instance( DT_ELECTRIC, rng( 1, 1 ) ) ).total_damage() > 0) {
+                sounds::sound(x, y, 30, _("You hear a crackle of electricity."));
+                if (g->u.sees( *this )){
+                    add_msg(_("The %s is shocked by the %1$s."), name().c_str(),
+                            ter_or_furn ? g->m.tername(x, y).c_str() : g->m.furnname(x, y).c_str());
```
Comment (BevapDin) [inline src/monmove.cpp]: Similar about `ter_or_furn` here.

File: src/mtype.h [hunk 5]
```diff
@@ -131,6 +131,7 @@ enum m_flag {
     MF_GROUP_BASH,          // Monsters that can pile up against obstacles and add their strength together to break them.
     MF_SWARMS,              // Monsters that like to group together and form loose packs
     MF_GROUP_MORALE,        // Monsters that are more courageous when near friends
+    MF_CLIMBS,              // Monsters that can climb certain terrain and furniture
```
Comment (BevapDin) [inline src/mtype.h]: You also need to add a line to `MonsterGenerator::init_flags()` to make this work.

File: src/npcmove.cpp [hunk 6]
```diff
@@ -1117,11 +1117,33 @@ void npc::move_to(int x, int y)
...
-            moves -= int(weapon.is_null() ? 80 : weapon.attack_time() * 0.8);;
-            int smashskill = str_cur + weapon.type->melee_dam;
-            g->m.bash( x, y, smashskill );
+        } else if (not g->m.has_flag_ter_or_furn( "ELECTRIFIED", x, y )) {
+        bool ter_or_furn = g->m.has_flag_ter_or_furn( "CLIMBABLE", x, y );
+            if (ter_or_furn) {
+            bool u_see_me = g->u.sees( *this );
+            int climb = dex_cur;
+                if (one_in( climb )) {
+                    if (u_see_me == true) {
```
Comment (BevapDin) [inline src/npcmove.cpp]: Why do people compare a boolean to another boolean. It is already a boolean, comparing it to true is pointless. Just write `if(u_see_me)`

**Comment by Rivet-the-Zombie** [comment]
Very nice!

File: src/game.cpp (hunk 2 above)
Comment (Lap) [inline src/game.cpp]: I think this also helps identify a bug outside this PR in game.cpp. Line 11996 and the block immediately after this relating to sharp terrain/furniture.

```C++
        if (m.has_flag("ROUGH", x, y) && (!u.in_vehicle)) {
            bool ter_or_furn = m.has_flag_ter_or_furn( "ROUGH", x, y );
            if (one_in(5) && u.get_armor_bash(bp_foot_l) < rng(2, 5)) {
                add_msg(m_bad, _("You hurt your left foot on the %s!"),
                        ter_or_furn ? m.tername(x, y).c_str() : m.furnname(x, y).c_str() );
                u.deal_damage( nullptr, bp_foot_l, damage_instance( DT_CUT, 1 ) );
            }
            if (one_in(5) && u.get_armor_bash(bp_foot_r) < rng(2, 5)) {
                add_msg(m_bad, _("You hurt your right foot on the %s!"),
                        ter_or_furn ? m.tername(x, y).c_str() : m.furnname(x, y).c_str() );
                u.deal_damage( nullptr, bp_foot_l, damage_instance( DT_CUT, 1 ) );
            }
        }
```

ter_or_furn should be m.ter_has_flag here as well should it not?

File: src/mtype.h (hunk 5 above)
Comment (Lap) [inline src/mtype.h]: Redid my whole branch after screwing up with git. Will readd.

**Comment by Coolthulhu** [comment]
As for the last point:
You could have jumper cables transform into a different part (not "connected jumper cable", but "jumper cable connected to fence") when the other part is connected to a fence.
While connected to fence, they wouldn't really work the same way as normal jumper cables do.

The biggest problem I see here would be keeping "in touch" with the electrifier vehicle. There is no good way of storing pointers to vehicles in terrain. You could probably store vehicle's position as tile's `cosmetic`.

You'd have to ensure the submap containing the powering vehicle is loaded, probably using `vehicle::find_vehicle` (this method is not `static`, but could easily be made `static`).

File: src/game.cpp (hunk 3 above)
Comment (Lap) [inline src/game.cpp]: Placeholder damage for now. I could make it hurt both hands/arms, one arm randomly, chest and arms, or the whole body. Any other suggestions?

 I was not aware of conductivity with weapons. I'll add that.

File: src/game.cpp (hunk 2 above)
Comment (BevapDin) [inline src/game.cpp]: Yes, that's a bug right there.

Comment (Lap) [inline src/game.cpp]: Well that's what I get for copy pasting. I'll fix it and the other one I saw in game.cpp as part of this PR I guess.

**Comment by Lap** [comment]
I'll have to add support for the fliers going over fences and avoiding getting electrified.

The damage type is DT_ELECTRIC. Shouldn't that cause a mitigation automatically? The stun part may have to be modified based on resistance.

File: src/game.cpp (hunk 3 above)
Comment (BevapDin) [inline src/game.cpp]: > Any other suggestions?

That's probably a bit complicated anyway. The zapback defense unconditionally damages the arms, that might be a good simple solution.
Another solution could be (not tested, identifiers are probably misspelled):
```C++
static const std::array<body_part, 6> suitable_bps = { bp_r_foot, bp_l_foot, ... };
const body_part target_bp = suitable_bps[rng(0, suitable_bps.size() - 1)];
```
But that gives each body part the same chance.

File: src/game.cpp (hunk 2 above)
Comment (narc0tiq) [inline src/game.cpp]: If you do that fix, do it in a different PR. Branches are cheap, sanity is not overrated.

File: src/monmove.cpp (hunk 1 above)
Comment (KA101) [inline src/monmove.cpp]: IIRC the numbering is there to help translate for languages with non-English word order.

**Comment by KA101** [comment]
You will need at least one electrified terrain object in order for this to merge--player-electrified fences would qualify.

This will be very popular on the forums.  Looking forward to it.

**Comment by DavidKeaton** [comment]
looking forward to this as well!

**Comment by Lap** [comment]
Hmmm...before I go tearing up more code with this one I'm at an impasse. 

map::has_flag has two functions:


```C++
bool map::has_flag(const std::string &flag, const int x, const int y) const
{
    static const std::string flag_str_REDUCE_SCENT("REDUCE_SCENT"); // construct once per runtime, slash delay 90%
    if (!INBOUNDS(x, y)) {
        return false;
    }

    int vpart;
    const vehicle *veh = veh_at( x, y, vpart );
    if( veh != nullptr && vpart.has_flag( flag ) && veh->obstacle_at_part( vpart ) >= 0 ) {
        return true;
    }

    return has_flag_ter_or_furn(flag, x, y);
}

bool map::has_flag(const ter_bitflags flag, const int x, const int y) const
{
    if (!INBOUNDS(x, y)) {
        return false;
    }

    int vpart;
    const vehicle *veh = veh_at( x, y, vpart );
    if( veh != nullptr && flag == TFLAG_REDUCE_SCENT && veh->obstacle_at_part( vpart ) >= 0 ) {
        return true;
    }

    return has_flag_ter_or_furn(flag, x, y);
}
```

Why are there two and why do either of them reference reduce_scent? Neither of them seem to function as I would expect (checking first if a vehicle on a location has a flag, then checking terrain and furniture).

File: src/monmove.cpp (hunk 1 above)
Comment (narc0tiq) [inline src/monmove.cpp]: When building without localization, the numbering is stripped away (for compatibility with some broken platform or other). For this reason, translatable strings must always be written as if the numbers will be sequential, and must have exactly as many values as formatting markers. Translated strings have more leeway -- building with gettext on the aforementioned broken platform is not supported -- so their numbers need not be sequential.

Also, to reiterate (and hopefully clarify), formatting indices start at 1, not 0, therefore `"the %1$s"` is a reference to the first value given to `add_msg`, not the second. The intent here seems to be either `"The %1$s climbs over the %2$s."`, or just simply `"The %s climbs over the %s."`. Personally, I'd do as @BevapDin suggests and go with the latter.

**Comment by narc0tiq** [comment]
@Lap Your first `has_flag` looks different from [the currently live master](https://github.com/CleverRaven/Cataclysm-DDA/blob/3edda48b25cebfcccef8a2670a3ade0efe4c7520/src/map.cpp#L1331) -- I don't know who changed your `flag_str_REDUCE_SCENT == flag` into a `vpart.has_flag(flag)`, but it's wrong.

Terrain flags and vehicle part flags are unrelated; the only place they come together is _if_ you're checking for the **terrain** flag `"REDUCE_SCENT"`, _and_ there's a vehicle obstacle part there, _then_ we short-circuit the checking of the terrain itself (via `has_flag_ter_or_furn`): the vehicle obstacle part counts as a temporary "REDUCE_SCENT" to the terrain under it.

As for why there are two of them? Because one takes an std::string parameter (freeform flag name) and the other takes a member of the `ter_bitflags` enum. See also [`src/mapdata.h:130-136`](https://github.com/CleverRaven/Cataclysm-DDA/blob/3edda48b25cebfcccef8a2670a3ade0efe4c7520/src/mapdata.h#L130), reproduced below:

```C++
 /*
  * Note; All flags are defined as strings dynamically in data/json/terrain.json and furniture.json. The list above
  * represent the common builtins. The enum below is an alternative means of fast-access, for those flags that are checked
  * so much that strings produce a significant performance penalty. The following are equivalent:
  *  m->has_flag("FLAMMABLE");     //
  *  m->has_flag(TFLAG_FLAMMABLE); // ~ 20 x faster than the above, ( 2.5 x faster if the above uses static const std::string str_flammable("FLAMMABLE");
  * To add a new ter_bitflag, add below and add to init_ter_bitflags_map() in mapdata.cpp
  * Order does not matter.
  */
```

**Comment by ejseto** [comment]
Electrified fences should be climbable if you're smart enough to never be touching both the fence and the ground at the same time. Maybe add an electronics or int check to see if you passed high school physics.

**Comment by KA101** [comment]
>Electrified fences should be climbable if you're smart enough to never be touching both the fence and the ground at the same time. Maybe add an electronics or int check to see if you passed high school physics.

Or PE to tell that it's electrified.  The one I encountered IRL wasn't marked, and I was told it wasn't electrified.  My cousins got a big laugh out of that one.  >_<

**Comment by kevingranade** [comment]
I think there's a bad assumption in there that the fence is all-or-nothing
electrified, I think fences targeting people alternate charged and grounded
sections for this reason.

**Comment by ejseto** [comment]
Then you would be able to climb it by touching only the grounded sections or only the charged sections :P

**Comment by Rivet-the-Zombie** [comment]
>Then you would be able to climb it by touching only the grounded sections or only the charged sections :P

Everything about this sounds like a gloriously bad idea.

**Comment by ejseto** [comment]
In real life? Sure. In a roguelike? Maybe I roll the dice.

**Comment by KA101** [comment]
OK, we've got stable, we've got major updates, can we get this?

**Comment by Lap** [comment]
I was out of town for a couple weeks of conference/vacation and haven't worked on this since. I'll see if I can get this done this week, though the final step with jumper cables may resist my abilities.

**Comment by KA101** [comment]
OK, hoping both were informative and rewarding, and let us know if we can help with the PR.

**Comment by Lap** [comment]
Will likely wait for #11895 before proceeding.

**Comment by Lap** [comment]
I stashed about 2/3 of these changes since the electric stuff still needed tweaking and all fences will need to be merged as the walls were done in #11895. 

This smaller PR should be ready to go.

**Comment by sashakiller33** [comment]
Hope soon will be electric lamps.
